### DISCUSSION TITLE
repeat action looping forever if not cleared

### DISCUSSION THREAD

**Discussion by herobank110** [excaliburjs/Excalibur#1852] [Q&A]
not sure if this is a bug or feature but my truck has an easeTo action followed by a callMethod. in this method I schedule some more actions with a repeat of (lets say) 3. I expected only those items just scheduled to be repeated, 3 times.

firstly, it repeated the 2nd round of actions forever which is surely a bug
secondly when using a `this.actions.clearActions()` before the 2nd round of actions it worked, but the actions happened N + 1 (in my case 3 + 1 = 4) times. maybe this is intentional since I guess the first time round is not 'repeated' and subsequently they are repeats, but if you look at semantic meaning of `repeat` eg in CSS or LOGO it is instead doing the action N times. if I had an `items.length - 1`, to catch the case of 1 length where essentially I would want the loop body once (no loop), it would repeat forever because zero = infinity
thirdly, repeat doesn't seem to respect delay actions in the loop body

**Comment by herobank110** [comment]
additionally, I have found callMethod action to be unreliable as it is sometimes not even called

**Reply by eonarheim** [reply]
@herobank110 

If you have a code sample where you can demonstrate the unreliable `callMethod` in a github issue and the repeat forever in a separate github issue that would help a ton.

As for actions in general, we do have it on our list to overhaul the internals of them (and possibly the api) because they have some subtle bugs/unexpected behaviors lurking (especially with the repeatX things). The internal implementation of repeat uses everything in the current queue to create it's list of things to repeat, and it relies on each action keeping track and resetting state properly for each repeat (which has proven to be a bad design decision)

I definitely agree with your semantic meaning of `repeat` that would be a better experience, and would be in line with CSS/LOGO expectations. I would support a change to this effect.

**Reply by herobank110** [reply]
I made an issue for the repeat forever but realised unreliable callMethod was actually my mistake in random action scheduling.
